### ISSUE TITLE
Error while deploying client

### ISSUE THREAD

**Issue by sitemapxml** [bacularis/bacularis-app#33]
I have installed and configured Bacula and Bacularis successfully, and everything seems to work properly.

However, there are errors when I try to deploy the client to Ubuntu or Debian (I haven't tested other systems, I am using only these two for servers).

After the Bacularis API host is successfully deployed ([like in this video](https://www.youtube.com/watch?v=wsayzd7CsuU)), I go to "Install Bacula components on API host" and try to install client (file daemon), and the following error appears in the "Raw output" tab (both server and client use Debian 12):

```
sudo /usr/bin/apt -y install bacula-fd 2>&1,,WARNING: apt does not have a stable CLI interface. Use with caution in scripts.,,Reading package lists...,Building dependency tree...,Reading state information...,The following additional packages will be installed:,  bacula-common,Suggested packages:,  gdb,The following NEW packages will be installed:,  bacula-common bacula-fd,0 upgraded, 2 newly installed, 0 to remove and 26 not upgraded.,Need to get 935 kB of archives.,After this operation, 2562 kB of additional disk space will be used.,Get:1 http://deb.debian.org/debian bookworm/main amd64 bacula-common amd64 9.6.7-7 [816 kB],Get:2 http://deb.debian.org/debian bookworm/main amd64 bacula-fd amd64 9.6.7-7 [119 kB],debconf: unable to initialize frontend: Dialog,debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.),debconf: falling back to frontend: Readline,debconf: unable to initialize frontend: Readline,debconf: (This frontend requires a controlling tty.),debconf: falling back to frontend: Teletype,dpkg-preconfigure: unable to re-open stdin:,Fetched 935 kB in 0s (56.8 MB/s),Selecting previously unselected package bacula-common.,(Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 52076 files and directories currently installed.),Preparing to unpack .../bacula-common_9.6.7-7_amd64.deb ...,Unpacking bacula-common (9.6.7-7) ...,Selecting previously unselected package bacula-fd.,Preparing to unpack .../bacula-fd_9.6.7-7_amd64.deb ...,Unpacking bacula-fd (9.6.7-7) ...,Setting up bacula-common (9.6.7-7) ...,debconf: unable to initialize frontend: Dialog,debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.),debconf: falling back to frontend: Readline,debconf: unable to initialize frontend: Readline,debconf: (This frontend requires a controlling tty.),debconf: falling back to frontend: Teletype,Setting up bacula-fd (9.6.7-7) ...,debconf: unable to initialize frontend: Dialog,debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.),debconf: falling back to frontend: Readline,debconf: unable to initialize frontend: Readline,debconf: (This frontend requires a controlling tty.),debconf: falling back to frontend: Teletype,,Creating config file /etc/default/bacula-fd with new version,debconf: unable to initialize frontend: Dialog,debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.),debconf: falling back to frontend: Readline,debconf: unable to initialize frontend: Readline,debconf: (This frontend requires a controlling tty.),debconf: falling back to frontend: Teletype,,Creating config file /etc/bacula/bacula-fd.conf with new version,Created symlink /etc/systemd/system/multi-user.target.wants/bacula-fd.service → /lib/systemd/system/bacula-fd.service.,Processing triggers for man-db (2.11.2-2) ...,sudo /usr/share/bacularis/protected/tools/set_bacula_perm.sh /etc/bacula www-data 2>&1,Internal error. [Warning] file_put_contents(/opt/bacula/etc/bacula-dir.conf): Failed to open stream: Permission denied (@line 64 in file /opt/bacularis-app/protected/vendor/bacularis/bacularis-common/Common/Modules/ConfigBacula.php).
```

There was also a second attempt to deploy the client on yet another Debian 12 instance and on Ubuntu 22.04, with similar (but not the same) output. These logs are attached as files.

On Ubuntu, there was a slightly different error which was not present when try to deploy to Debian, and that was the following part of the log:

```
Processing triggers for systemd (245.4-4ubuntu3.22) ...,sudo /usr/share/bacularis/protected/tools/set_bacula_perm.sh /etc/bacula www-data 2>&1,Config validation error.Array
(
    [output] => JSON tool returned wrong exitcode.
 Output:sudo: /usr/sbin/bfdjson: command not found
    [exitcode] => 82
)
```

Here are the additional logs of another two attempts:
[Bacularis-error2-debian-12.txt](https://github.com/bacularis/bacularis-app/files/14394651/Bacularis-error2-debian-12.txt)
[Bacularis-error3-ubuntu-20.04.txt](https://github.com/bacularis/bacularis-app/files/14394652/Bacularis-error3-ubuntu-20.04.txt)

@ganiuszka Do you have some idea what might be causing the problem?

**Comment by sitemapxml** [comment]
Just to add, the bacula was installed from the official repository on bacula.org (bacula 13.04 was installed on the server), and it seems that it use different path for `bfdjson` tool than version 9.XX that is in official Debian repositories.
The old path was `/usr/sbin/bfdjson`, and now it has been installed to `/opt/bacula/bin/bfdjson`.

Looking at the Ubuntu related error log, this is probably the part of the issue, but still I can't figure out what is the problem with Debian deploy.

**Comment by ganiuszka** [comment]
Hello @sitemapxml,

Thank you for your detailed report.

The video guide that you pointed is for deploying Bacularis and Bacula using the system Debian package repositories. For using bacula.org repository you need to use this video guide:

https://www.youtube.com/watch?v=EkSceivoUyc

The key here to solve your problem with permissions is to create a new OS profiles for Debian and Ubuntu that base on the original Debian and Ubuntu OS profiles and that have paths adapted to the Bacula from bacula.org packages. For example one of things to change is the parameter in a script to set permissions that we can see in the output:

```
sudo /usr/share/bacularis/protected/tools/set_bacula_perm.sh /etc/bacula www-data 2>&1
```

there should be:

```
sudo /usr/share/bacularis/protected/tools/set_bacula_perm.sh /opt/bacula/etc www-data 2>&1
```

Also some Bacula package names have different name between the system Debian/Ubuntu repositories and bacula.org repository.

In other words, the customized OS profiles are answer to your first problem. You will see on the video guide that I pointed.

For the Ubuntu 20.04 with installation by system packages and the error about bfdjson it is a bug with wrong default value in the OS profile for Ubuntu 20.04 because on this system the ``bfdjson`` path is:

```
/usr/lib/bacula/bfdjson
```
As a workaround you can change this path in the customzied OS profile that base on the original. Targetly I will fix it at the moment. Thanks for catching it.

Best regards,
Marcin Haba (gani)

**Comment by ganiuszka** [comment]
Hello @sitemapxml

I fixed this problem with b*json paths in OS profile for Ubuntu 20.04.

I think we can close this issue. Good luck with the Bacula deploying. In case other problems with this Bacularis function, please let us know.

Thanks.

Best regards,
Marcin Haba (gani)

**Comment by sitemapxml** [comment]
Hello, Thank you for your response, and thanks for the fix.

I did as per video you linked (on the fresh Ubuntu 22.04 instance), and so far, there is some progress, although, I still didn't manage to install the client after deploying the API host.

The new error is as follows:
```
sudo /usr/share/bacularis/protected/tools/set_bacula_perm.sh /opt/bacula/etc www-data 2>&1,Internal error. [Warning] file_put_contents(/opt/bacula/etc/bacula-dir.conf): 
Failed to open stream: Permission denied (@line 64 in file /opt/bacularis-app/protected/vendor/bacularis/bacularis-common/Common/Modules/ConfigBacula.php).
```

And this is the full log (after the part with "Reading database ..."):
```
Preparing to unpack .../bacula-client_13.0.4-24021217~jammy_amd64.deb ...,
Unpacking bacula-client (13.0.4-24021217~jammy) ...,
Setting up bacula-client (13.0.4-24021217~jammy) ...,
debconf: unable to initialize frontend: Dialog,debconf: (Dialog frontend will not work on a dumb terminal, an emacs shell buffer, or without a controlling terminal.),
debconf: falling back to frontend: Readline,Created symlink /etc/systemd/system/multi-user.target.wants/bacula-fd.service → /lib/systemd/system/bacula-fd.service.,
Processing triggers for man-db (2.10.2-1) ...,,Running kernel seems to be up-to-date.,,
No services need to be restarted.,,No containers need to be restarted.,,No user sessions are running outdated binaries.,,
No VM guests are running outdated hypervisor (qemu) binaries on this host.,
sudo /usr/share/bacularis/protected/tools/set_bacula_perm.sh /opt/bacula/etc www-data 2>&1,Internal error. [Warning] file_put_contents(/opt/bacula/etc/bacula-dir.conf): 
Failed to open stream: Permission denied (@line 64 in file /opt/bacularis-app/protected/vendor/bacularis/bacularis-common/Common/Modules/ConfigBacula.php).
```

The owner of the files under `/opt/bacularis-app` was `root`, so I tried to change that to `www-data` user and group for all the files, but it didn't solve the problem.

**Comment by sitemapxml** [comment]
Just to let you know, I managed to solve this. The issue was that I forgot to set proper permissions for the bacula config directory, so the app could not write configuration to bacula files.

After setting proper ownership, everything works perfectly!

Thank you once again.

**Comment by ganiuszka** [comment]
Hello @sitemapxml

Thanks the the additional details about your tries with deploying. I am glad that you managed to solve the problem with Ubuntu 22.04.

By the way, you don't need to set permissions for the Bacula config directory manually. It is done by the ``set_bacula_perm.sh`` script.

One thing that I am not sure, when I am looking at the outputs, is about the directory where the Bacularis is installed. It looks that it is in:

```
/opt/bacularis-app
```

but the ``set_bacula_perm.sh`` path uses (which is typical for installation using Bacularis binary packages):

```
/usr/share/bacularis
```

So, I think that you need to adapt a bit the ``set_bacula_perm.sh`` script path in your OS profile:

```
/usr/share/bacularis/protected/tools/set_bacula_perm.sh /opt/bacula/etc www-data 2>&1
```

It is in the post-install actions in advanced setting of the OS profile for every Bacula components.

Good luck.

Best regards,
Marcin Haba (gani)

**Comment by sitemapxml** [comment]
Ah, sorry... I forgot to mention that Bacularis was installed using composer, and in my case, it is installed inside `/opt/bacularis-app` directory. Will definitely use the script for the next installation.
